More Rust guidelines (#7360)
Azure/azure-sdk: docs/rust/introduction.md

{% include requirement/MUST id="rust-general-idiomatic-code" %} write idiomatic Rust code. If you’re not familiar with the language, a great place to start is <https://www.rust-lang.org/learn>. Do __NOT__ simply attempt to translate your language of choice into Rust.

{% include requirement/MUSTNOT id="rust-general-version" %} use grammar or features newer than the `rust-version` declared in the root `Cargo.toml` workspace.

{% include requirement/MUSTNOT id="rust-general-dependencies" %} necessarily depend on any particular async runtime or HTTP stack.

{% include requirement/MUSTNOT id="rust-general-unwrap" %} call `unwrap()`, `expect()`, or other functions that may panic unless you are absolutely sure they never will. It's almost always better to use `map()`, `unwrap_or_else()`, or a myriad of related functions to remap errors, return suitable defaults, etc.

> {% include requirement/SHOULD id="rust-general-dependencies-default" %} depend on `tokio` (async runtime) and `reqwest` (HTTP stack) in the `default` feature for crates.

Author: We're already doing it in the existing code base, but I can walk you through it.

The keyword in line 67 is "necessarily", but I can see how others might interpret that in the common vernacular. I'll reword.

> {% include requirement/SHOULD id="rust-general-dependencies-default" %} depend on `tokio` (async runtime) and `reqwest` (HTTP stack) in the `default` feature for crates.

Author: Reordering to put the bullets together might help a little as well.

{% include requirement/MUSTNOT id="rust-client-methods-async-api" %} provide a synchronous programming model for service methods.

##### Naming

{% include requirement/MUST id="rust-client-methods-naming-case" %} use snake_case method names converted from likely either camelCase or PascalCase declared in the service specification e.g., `getResource` would be declared as `get_resource`.

{% include requirement/MUST id="rust-client-methods-naming-list" %} use the `list_` prefix for service methods that return one or more pages containing a list of resources e.g., `list_properties_of_secrets()`.

{% include requirement/MUST id="rust-client-methods-naming-conversion-prefix" %} use the following prefixes in the described scenarios:

| Prefix | Scenario | Example |
| ------ | -------- | ------- |
| (none) | field getter | `field_name(&self) -> FieldType` |
| `with_` | field setter returning `Self` - typically used in builders | `with_field_name(&mut self, value: FieldType) -> &mut Self` |
| `set_` | field setter returning nothing or anything else | `set_field_name(&mut self, value: FieldType)` |

#### Service Method Parameters {#rust-parameters}

> {% include requirement/MUST id="rust-parameters-self" %} take a `&self` as the first parameter. All service clients must be immutable

Reviewer: ```suggestion
{% include requirement/MUST id="rust-parameters-self" %} take a `&self` as the first parameter. All service clients must be threadsafe
```

Reviewer: But, you have said this a few times already so I dont' know if you need to say this again.

##### Parameter Validation

The service client will have several methods that perform requests on the service. _Service parameters_ are directly passed across the wire to an Azure service. _Client parameters_ are not passed directly to the service, but used within the client library to fulfill the request. Examples of client parameters include values that are used to construct a URI or a file that needs to be uploaded to storage.

{% include requirement/MUST id="rust-parameters-client-validation" %} validate client parameters.

{% include requirement/MUSTNOT id="rust-parameters-server-validation" %} validate service parameters. This includes null checks, empty strings, and other common validating conditions. Let the service validate any request parameters.

{% include requirement/MUSTNOT id="rust-parameters-server-defaults" %} encode service parameter default values. These values may change from version to version and may cause unexpected results when calling older versions from a newer client. Let the service apply default parameter values.

> {% include requirement/MUST id="rust-parameters-check-devex" %} validate the developer experience when the service parameters are invalid to ensure appropriate error messages are generated by the service. If the developer experience is compromised due to service-side error messages, work with the service team to correct prior to release.

Reviewer: I like this guideline! Very few people on the SDK team adhere to this and I wish they would.

##### Operation Options

> TODO: Pass per-call options that affect the pipeline e.g., if Some(T), clone pipeline and modify.

> ##### Return Types

Reviewer: I didn't see a guideline on how to structure a full client method (not exact pseudo code):
Result<Response<ResourceModel>> SomeOperation(Mandatory1, Mandatory2, ResourceModel, SomeOperationOptions)

Author: Added above in the Service Methods section opener.

The API surface of your client library must have the most thought as it is the primary interaction that the consumer has with your service.

{% include requirement/MUST id="rust-api-naming-concise" %} use clear, concise, and meaningful names.

{% include requirement/MUST id="rust-api-naming-standard" %} follow [Rust naming conventions][rust-lang-naming].

> {% include requirement/SHOULDNOT id="rust-api-naming-abbreviation" %} use abbreviations unless necessary or when they are commonly used and understood. For example, `iot` is used since it is a commonly understood industry term; however, using `kv` for Key Vault would not be allowed since `kv` is not commonly used to refer to Key Vault.

Author: Good call out. Module and method names will always be `iot` but looking at a few popular IoT crates, type names seems to use `Iot` more commonly (almost completely; the one `IoT` in one of them seemed like an oversight).

### Service Client {#rust-client}

Service clients are the main starting points for developers calling Azure services with the Azure SDK. Each client library should have at least one client in its main namespace, so it’s easy to discover. The guidelines in this section describe patterns for the design of a service client.

There exists a distinction that must be made clear with service clients: not all classes that perform HTTP (or otherwise) requests to a service are automatically designated as a service client. A service client designation is only applied to classes that are able to be directly constructed because they are uniquely represented on the service. Additionally, a service client designation is only applied if there is a specific scenario that applies where the direct creation of the client is appropriate. If a resource can not be uniquely identified or there is no need for direct creation of the type, then the service client designation should not apply.

{% include requirement/MUST id="rust-client-name" %} name service client types with the _Client_ suffix e.g., `SecretClient`.

> Client names are specific to the service to avoid ambiguity when using multiple clients without requiring `as` to change the binding name e.g.,

Author: Clarified.

  --> src/main.rs:2:38
   |
1  |     use azure_storage_blob::Client;
   |         ----------- previous import of the type `Client` here
2  |     use azure_security_keyvault::secrets::Client;
   |         ^^^^^^^^^^^ `Client` reimported here
   |
```

{% include requirement/SHOULD id="rust-client-namespace" %} place service client types that the consumer is most likely to interact with in the root module of the client library e.g., `azure_security_keyvault`. Specialized service clients should be placed in submodules e.g., `azure_security_keyvault::secrets`.

{% include requirement/MUST id="rust-client-immutable" %} ensure that all service client classes thread safe (usually by making them immutable and stateless).

{% include requirement/MUST id="rust-client-endpoint" %} define a public `endpoint(&self) -> &azure_core::Url` method to get the endpoint used to create the client.

#### Service Client Constructors {#rust-client-constructors}

{% include requirement/MUST id="rust-client-constructors-new" %} define a public function `new` that takes the following form and returns `Self` or `azure_core::Result<Self>` if the function may fail.

```rust
impl SecretClient {

>     pub fn new(endpoint: impl AsRef<str>, credential: std::sync::Arc<dyn azure_core::TokenCredential>, options: Option<SecretClientOptions>) -> azure_core::Result<Self> {

Author: Not all TokenCredentials are related to AAD, but you're saying you have some credential types that don't implement TokenCredential? We might carve out specific construction patterns borrowing from guidelines elsewhere, like either in addition to or in lieu of `new()` they could also have `with_shared_key_credential()` or something like that. This (with `with_` or `from_` prefixes) is a fairly common pattern I see in other crates. `new()` is confidered the baseline or recommended pattern, but maybe that doesn't always work in cases where clients (a few come to mind) don't necessarily recommend one over the other...though, aren't we actually recommending AAD over shared key for security these days? So maybe `new()` still works in these scenarios.

>     pub fn new(endpoint: impl AsRef<str>, credential: std::sync::Arc<dyn azure_core::TokenCredential>, options: Option<SecretClientOptions>) -> azure_core::Result<Self> {

Author: Added some text about this.

        let endpoint = azure_core::Url::parse(endpoint.as_ref())?;
        let options = options.unwrap_or_default();
        todo!()
    }
}
```

{% include requirement/MAY id="rust-client-constructors-credential" %} accept a different credential type if the service does not support AAD authentication.

##### Client Configuration {#rust-client-configuration}

{% include requirement/MUST id="rust-client-configuration-name" %} name the client options struct with the same as the client name + "Options" e.g., a `SecretClient` takes a `SecretClientOptions`.

> {% include requirement/MUST id="rust-client-configuration-version" %} define a `pub api_version: String` field to pass to the service for all HTTP clients.

Author: Clarified. I meant it must be defined for all HTTP clients, but I understand the confusion.

> {% include requirement/MUST id="rust-client-configuration-core" %} define a `pub options: azure_core::ClientOptions` field to define global configuration shared by all HTTP clients.

Author: Clarified a bit.

{% include requirement/MUST id="rust-client-configuration-clone" %} derive `Clone` to support cloning client configuration for other clients.

{% include requirement/MUST id="rust-client-configuration-default" %} derive `Default` to support creating default client configuration including the default `api-version` used when calling into the service.

The requirements above would define an example client options struct like:

```rust
#[derive(Clone, Default)]
pub struct SecretClientOptions {
    pub api_version: String,
    pub options: azure_core::ClientOptions,
}
```

> {% include requirement/MUSTNOT id="rust-client-configuration-env" %} use client library-specific runtime configuration such as environment variables or configuration files. Some environments e.g., WASM or many IoT devices won't have access to an environment block or file system.

Reviewer: Great point! I would not have thought of this myself.

{% include requirement/MUSTNOT id="rust-client-configuration-env-sys" %} change the default values of the client options based on system or program state.

{% include requirement/MUSTNOT id="rust-client-configuration-env-build" %} change the default values of the client options based on how the client library was built.

{% include requirement/MUSTNOT id="rust-client-configuration-immutable" %} change the behavior of the client after the client is constructed with the following exceptions:

* Log level, which must take effect immediately across all client libraries.
* Tracing on or off, which must take effect immediately across all client libraries.

##### Service Versions {#rust-client-api-version}

{% include requirement/MUST id="rust-client-api-version-latest" %} call the latest supported service API version by default. Typically this will be the API version from which the client library was generated.

> {% include requirement/MUST id="rust-client-api-version-override" %} allow the consumer to explicitly select a supported service API version when instantiating the service client.

Reviewer: We don't check if the customer-passed api-version is supported by the service or not; the customer just passes any value and we attempt to use it and the chips fall where they may. We may not of any private preview version that shows up and an old SDK will not know if an old service version has been retired and just stops working entirely.

Author: Just to clarify - your comment is about the word "supported" as in "supported service API" here, right? I'll assume so, but please reopen with a comment if not.

##### Mocking {#rust-client-mocking}

{% include requirement/MUST id="rust-client-mocking-trait-name" %} define a trait named after the client name + "Methods" e.g., `SecretClientMethods`.

> {% include requirement/MUST id="rust-client-mocking-trait-methods" %} implement all methods of the client methods trait on the client which have the body `unimplemented!()` or `std::future::ready(unimplemented!())` for async methods e.g.,

Reviewer: Isn't this brittle? If a new version of the client adds new methods, then this type will get new methods as well and won't the customer HAVE TO implement those methods to get things to compile again? At some point, you can walk me through mocking but it is low priority for now.

Author: No. Trait methods that have a body use the body by default - almost like virtual methods in some other languages. When you impement trait methods, you effectively override them. Tons of std traits - especially those related to iteration - do this.

> TODO: This section needs to be driven by code in the Core library.

##### Cancellation

Cancelling an asynchronous method call in Rust is done by dropping the `Future`.

The Rust `std` crate itself does not implement an async runtime, so different async runtimes must be chosen by the caller and may support cancellation different. `tokio` is common and should be the default, but not required.
Various extensions also exist that the caller may use that may otherwise not work with passing in a cancellation token like in some other Azure SDK languages.

##### Thread Safety

> {% include requirement/MUST id="rust-client-methods-thread-safety" %} be thread-safe. All public members of the client type must be safe to call from multiple threads concurrently.

Reviewer: This has been said multiple times before

Author: A lot of guidelines in other languages are repetitive. Most of this structure is from the Go guidelines. I will try to reduce, though. I already did for doc comments.

If you have cause to release separate crates for a single service, please discuss first with the [Architecture Board].

{% include requirement/SHOULD id="directory-structure-lib" %} only export public APIs from the crate `lib.rs` and define all other types in suitable [modules][rust-lang-modules]:

* Single-file modules should be declared in a file next to their parent module.
* Multi-file modules should be declared in a directory next to their parent module with a `mod.rs` file.

For example:

```text
src/
  policies/
    client_id.rs
    mod.rs
    retry.rs
    transport.rs
  error.rs
  lib.rs
Cargo.lock
Cargo.toml
```

#### Common Libraries

{% include requirement/MUST id="rust-common-macros-review" %} review new macros with your language architect(s).

> {% include requirement/MAY id="rust-common-macros-core" %} use common procedural macros from `azure_core`.

Reviewer 2: I assume this requires us to make those MACROS public and thus the public could also use them? Do we want that?

Author: Yes, they need to be public but we only support them within our crates. Those macros most often make use of types declared by our crates, so they're of little use in unsupported scenarios so I don't see why we'd be concerned.

Reviewer 2: If they are public, then anyone can use. Documenting is difficult to enforce. I don't have a solution that is better.

It would be great if we had a way to partition certain types as for use by AzureSDK authors only, vs types that end-user developers should use. Maybe we namesepace em or put em in a different crate/module?????

Author: Anyone could use any types from any of our libraries in unsupported ways. I don't understand the problem.

Reviewer 2: Agreed. We are OSS and MIT license so they can copy at will as well. Them copying or doing strange things is something we can't stop. We don't imply any support in those cases.  

We don't however have to embrace publicly shipping/supporting a MACRO. A publicly exposed item has an implied level of support.

Author: > As example of this is the internal folder that .NET copies around and compiles into each azure sdk library such that the libraries have the access to shared internal code.

We didn't copy it. Project files can refer to source outside of its parent directory. That's not the case here. Once we have a stable API - all APIs - I wouldn't expect to change macros (no need to keep capitalizing it :wink:) any more than we would any other API. This is no different than the oodles of macros every popular rust crate - including `std` itself - do. Most, if not all, macros will likely emit `azure_core`-specific types anyway, so wouldn't generally be useful outside of our use cases.

### Versioning {#rust-versioning}

#### Client Versions

{% include requirement/MUST id="rust-versioning-backwards-compatibility" %} be 100% backwards compatible with older versions of the same package.

> {% include requirement/MUST id="rust-versioning-major-version" %} increase the major semantic version number if an API breaking change is required.

Reviewer 2: This comment makes sense. Any reason to NOT also specify following semver?

- https://semver.org/

> Breaking changes should happen rarely, if ever. Register your intent to do a breaking change with the [Architecture Board]. You'll need to have a discussion with the language architect before approval.

Reviewer 2: ```suggestion
Breaking changes should happen rarely, if ever. Register your intent to do a breaking change with the [Architecture Board].
{% include requirement/MUST id="rust-versioning-major-version-archreview" %} discuss with the language architect before proceeding with a breaking change.
```

> Breaking changes should happen rarely, if ever. Register your intent to do a breaking change with the [Architecture Board]. You'll need to have a discussion with the language architect before approval.

Reviewer: I think of this as having 2 kinds of breaking changes:
1. SDK induced where the Rust SDK team decides to make some break for a really, really good reason.
2. Service induced where the Rust SDK team has no choice but to break the SDK package. 

I think the distinction should make it into the guidelines so that an SDK engineer doesn't feel like they are violating a guideline when the service is forcing the break.

Author: Thanks. I'll call this out and reword with [@RickWinter's comment above](https://github.com/Azure/azure-sdk/pull/7360#discussion_r1530703227) in mind.

Author: Done. Please reopen, @JeffreyRichter or @RickWinter, if you don't like the new wording. I waffled a bit over a few variations and still not 100% I like what I'll push soon, but we can always change it later...sooner or later.

Consistent version number scheme allows consumers to determine what to expect from a new version of the library.

{% include requirement/MUST id="rust-package-versions-semver" %} use _MAJOR_._MINOR_._PATCH_ format for the version of the library dll and the NuGet package.

Use _-beta._N_ suffix for beta package versions. For example, _1.0.0-beta.2_.

{% include requirement/MUST id="rust-package-versions-change-on-release" %} change the version number of the client library when __ANYTHING__ changes in the client library.

{% include requirement/MUST id="rust-package-versions-patching" %} increment the patch version when fixing a bug.

{% include requirement/MUSTNOT id="rust-package-versions-features-in-patch" %} include new APIs in a patch release.

{% include requirement/MUST id="rust-package-versions-add-feature" %} increment the major or minor version when adding support for a service API version.

{% include requirement/MUST id="rust-package-versions-add-api" %} increment the major or minor version when adding a new method to the public API.

{% include requirement/SHOULD id="rust-package-versions-major-changes" %} increment the major version when making large feature changes.

> {% include requirement/MUST id="rust-package-versions-change-on-release" %} select a version number greater than the highest version number of any other released Track 1 package for the service in any other scope or language.

Reviewer: Is this necessary for Rust? Are you considering our prior SDKs as Track 1 even though they never formally GA'd?

Author: We are considering that "track 1", yes, but not in official conversations. Maybe we could elide it.

Author: Actually, it does say "in any other scope or language". At least for KV we adhered to that. Is that normal for all packages across languages? I can clean up any mention of "track 1", though, but it still works in context.

Reviewer: Yes, I would elide. Plus it is moment in time and will become meaningless by end of year-ish.

Dependencies bring in many considerations that are often easily avoided by avoiding the dependency.

* __Versioning__ - Many programming languages do not allow a consumer to load multiple versions of the same package. So, if we have an client library that requires v3 of package Foo and the consumer wants to use v5 of package Foo, then the consumer cannot build their application. This means that client libraries should not have dependencies by default.
* __Size__ - Consumer applications must be able to deploy as fast as possible into the cloud and move in various ways across networks. Removing additional code (like dependencies) improves deployment performance.
* __Licensing__ - You must be conscious of the licensing restrictions of a dependency and often provide proper attribution and notices when using them.
* __Compatibility__ - Often times you do not control a dependency and it may choose to evolve in a direction that is incompatible with your original use.
* __Security__ - If a security vulnerability is discovered in a dependency, it may be difficult or time consuming to get the vulnerability corrected if Microsoft does not control the dependencies code base.

> {% include requirement/MUST id="rust-dependencies-centralized" %} declare all dependencies in the repository root `Cargo.toml` workspace in the `[dependencies]` section regardless of which type of dependency creates will inherit, e.g.:

Reviewer 2: ```suggestion
{% include requirement/MUST id="rust-dependencies-centralized" %} declare all dependencies in the repository root `Cargo.toml` workspace in the `[dependencies]` section regardless of which type of dependency crates will inherit, e.g.:
```

Reviewer 2: Myself included...

This will impact line numbers, so you should only export APIs publicly from `lib.rs`.

{% include requirement/MUST id="rust-documentation-parameters" %} document all parameters. Prior to [conventional doc comment markdown headers][rust-lang-rustdoc-headings], declare a `Arguments` heading as needed (not needed for `&self`):

```rust
/// Sets a secret and returns more information about the secret from the service.
///
/// # Arguments
///
/// * `name` - The name of the secret.
/// * `value` - The value of the secret.
/// * `options` - Optional properties of the secret or method call options.
fn set_secret(&self, name: impl Into<String>, value: impl Into<String>, options: Option<SetSecretOptions>) - Result<Response>;
```

See [Rust by Example: Documentation][rust-lang-doc-meta] for more information.

> {% include requirement/SHOULD id="rust-documentation-tests" %} use testable examples in documentation which not only improves test coverage, but also shows callers runnable examples.

Reviewer 2: NIT, readability.
```suggestion
{% include requirement/SHOULD id="rust-documentation-tests" %} use testable examples in documentation which improve test coverage and show callers runnable examples.
```
